tests/summ_ai_api/utils.py

```diff
@@ -0,0 +1,164 @@
+from unittest.mock import MagicMock, Mock
+
+import aiohttp
+from asgiref.sync import sync_to_async
+
+from integreat_cms.cms.constants import machine_translation_permissions as mt_perms
+from integreat_cms.cms.models import Page, Region
+
+
+def make_mock_summ_ai_server(response_data, status=200):
+    """
+    Creates a mock of SummAiServer-(Response)
+
+    :param response_data: data that mocked server returns
+    :type response_data: dict
+
+    :param status: http status code that mocked server returns (default = 200)
+    :type status: int
+
+    :return: Request handler for mocked SummAiServer
+    :rtype: ~collections.abc.Callable
+    """
+
+    async def fake_summ_ai_server(request):
+        """
+        Create fake responses which simulate the SUMM.AI API server
+
+        :param request: The request
+        :type request: aiohttp.web.Request
+
+        :return: The response
+        :rtype: aiohttp.web.Response
+        """
+        return aiohttp.web.json_response(data=response_data, status=status)
+
+    return fake_summ_ai_server
+
+
+def make_mock_summ_ai_server_rate_limited():
+    """
+    Creates a mock of SUMM.AI server with rate limiting
+
+    :return: Mocked SUMM.AI server with rate limiting
+    :rtype: aiohttp.web.Application
+    """
+    app = aiohttp.web.Application()
+    app["attempt"] = 0
+
+    async def fake_summ_ai_server_rate_limited(request):
+        request.app["attempt"] += 1
+        if request.app["attempt"] < 2:
+            # What if we get rate limited?
+            return aiohttp.web.json_response(
+                data={"error": "rate limit exceeded"}, status=429
+            )
+        if request.app["attempt"] == 2:
+            # What if we get invalid JSON?
+            return aiohttp.web.Response(
+                text='{"incomplete json response": ', status=200
+            )
+        return aiohttp.web.json_response(
+            data={
+                "translated_text": "Hier ist Ihre Leichte Sprache Übersetzung",
+                "jobid": "9999",
+            },
+            status=200,
+        )
+
+    app.router.add_post("/", fake_summ_ai_server_rate_limited)
+    return app
+
+
+def make_broken_json_summ_ai_server(response_text, status=200):
```

Comment:
This method is just the same as make_mock_summ_ai_server().
I guess we could use make_mock_summ_ai_server() as a universal one.

Reply:
It is not, this one allows us to inject any *text* into the request, while the other is more convenient as it takes a `dict()` and turns that into JSON. We need this separate instance because we test the behaviour on bogus data that cannot be parsed as JSON. However, just looking at this collection of functions in `utils.py`, this context is lost, the difference between the two functions is not obvious at all at first (or second) glance and I worry this might be a big source of confusion in the future. Do you think adding additional doctrings would be sufficient?

Reply:
@PeterNerlich what confused me here: 
This method is named as “broken” but it accepts any text and can basically accept a valid json too.